BrightID/src/components/Apps/appThunks.ts

```diff
@@ -0,0 +1,588 @@
+import { Alert } from 'react-native';
+import { create } from 'apisauce';
+import { t } from 'i18next';
+import { getSponsorship } from '@/components/Apps/model';
+import { getGlobalNodeApi } from '@/components/NodeApiGate';
+import {
+  addLinkedContext,
+  selectAllSigs,
+  selectAppInfoByAppId,
+  selectLinkingAppError,
+  selectLinkingAppInfo,
+  selectApplinkingStep,
+  setApps,
+  setLinkingAppError,
+  setLinkingAppInfo,
+  setLinkingAppStarttime,
+  setAppLinkingStep,
+  setSponsorOperationHash,
+  updateLinkedContext,
+  updateSig,
+} from '@/reducer/appsSlice';
+import {
+  BrightIdNetwork,
+  operation_states,
+  SPONSOR_WAIT_TIME,
+  SPONSORING_POLL_INTERVAL,
+  app_linking_steps,
+} from '@/utils/constants';
+import { addOperation, Operation } from '@/reducer/operationsSlice';
+import { NodeApi } from '@/api/brightId';
+import {
+  selectIsSponsored,
+  setIsSponsoredv6,
+  userSelector,
+} from '@/reducer/userSlice';
+import { selectIsPrimaryDevice, updateBlindSig } from '@/actions';
+
+type requestLinkingParams = {
+  appId: string;
+  appUserId: string;
+  baseUrl?: string;
+  v: number;
+};
+export const requestLinking =
+  (params: requestLinkingParams): AppThunk<Promise<void>> =>
+  async (dispatch: AppDispatch, getState) => {
+    const appLinkingStep = selectApplinkingStep(getState());
+    if (appLinkingStep !== app_linking_steps.IDLE) {
+      console.log(
+        `Can't request linking when not in IDLE state. Current state: ${appLinkingStep}`,
+      );
+      return;
+    }
+    const linkingError = selectLinkingAppError(getState());
+    if (linkingError) {
+      console.log(
+        `Can't request linking when there is still an active error. Current errir: ${linkingError}`,
+      );
+      return;
+    }
+
+    // store app linking details
+    dispatch(setLinkingAppInfo(params));
+
+    const api = getGlobalNodeApi();
+    dispatch(setAppLinkingStep({ step: app_linking_steps.REFRESHING_APPS }));
+    try {
+      // make sure to have latest appInfo available
+      const apps = await api.getApps();
+      dispatch(setApps(apps));
+    } catch (e) {
+      const msg = e instanceof Error ? e.message : `${e}`;
+      dispatch(setLinkingAppError(`Failed to fetch latest appInfo: ${msg}`));
```

Comment:
It seems if such problems happen, users should reset the app to be able to retry linking. yes?